rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+                    * (meta_data::size::PAGES_IN_BLOCK as u64)
+        } else {
+            // Before initialization of anything, give back zero.
+            0
+        }
+    };
+    size
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_set_mem_size<M: Memory>(_mem: &mut M, size: u64) {
+    crate::region::REGION_SET_MEM_SIZE = Some(size);
+}
+
+// Helper for commmon logic that reserves low-valued RegionIds in a certain span for future use.
+// When first is some, we are actually reserving.  When first is none, we are checking that the reservation has occured.
+unsafe fn region_reserve_id_span<M: Memory>(_mem: &mut M, first: Option<RegionId>, last: RegionId) {
+    if let Some(first) = first {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert_eq!(first.0, next_id);
+        assert!(first.0 <= last.0);
+        meta_data::total_allocated_regions::set((last.0 + 1) as u64);
+    } else {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert!(last.0 < next_id);
+    }
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
+    let next_id = meta_data::total_allocated_regions::get() as u16;
+    meta_data::total_allocated_regions::set(next_id as u64 + 1);
+
+    let vec_pages = alloc_blob(mem, Bytes(0));
+    let r_ptr = alloc_region(mem, next_id, 0, vec_pages);
+
+    // Update Region table.
+    {
+        let r_id = RegionId::from_id(next_id);
+        let c = meta_data::region_table::get(&r_id);
+        // sanity check: Region table says that this region is available.
+        assert_eq!(c, None);
+        meta_data::region_table::set(&r_id, Some(RegionSizeInPages(0)));
+    }
+    r_ptr
+}
+
+pub unsafe fn region_recover<M: Memory>(mem: &mut M, rid: &RegionId) -> Value {
+    use meta_data::size::PAGES_IN_BLOCK;
+    let c = meta_data::region_table::get(&rid);
+    let page_count = match c {
+        Some(RegionSizeInPages(pc)) => pc,
+        _ => {
+            rts_trap_with("region_recover failed: Expected Some(RegionSizeInPages(_))");
+        }
+    };
+
+    let block_count = (page_count as u32 + PAGES_IN_BLOCK - 1) / PAGES_IN_BLOCK;
+    let vec_pages = alloc_blob(mem, Bytes(block_count * 2));
+
+    let r_ptr = alloc_region(mem, rid.0, page_count as u32, vec_pages);
+
+    let tb = meta_data::total_allocated_blocks::get();
+    let av = AccessVector(vec_pages.as_blob_mut());
+    let mut recovered_blocks = 0;
+    for block_id in 0..tb {
+        match meta_data::block_region_table::get(BlockId(block_id as u16)) {
+            None => {}
+            Some((rid_, rank)) => {
+                if &rid_ == rid {
+                    av.set_ith_block_id(rank.into(), &BlockId(block_id as u16));
+                    recovered_blocks += 1;
+                }
+            }
+        }
+    }
+    assert_eq!(recovered_blocks, block_count);
+    r_ptr
```

Comment:
Allocation barrier required for the incremental GC:

```suggestion
    allocation_barrier(vec_pages);
    r_ptr
```